tests/tensorflow/test_printing.py

```diff
@@ -79,3 +79,90 @@ def test_unknown_shape():
+def test_shape_index():
+    """Make sure graph indexing functions as expected."""
+
+    A = tf.compat.v1.placeholder('float', name='A',
+                                 shape=tf.TensorShape([None, None]))
+    x = tf.compat.v1.placeholder('float', name='x',
+                                 shape=tf.TensorShape([None, 1]))
+    y = tf.multiply(1.0, x, name='y')
+
+    z = tf.matmul(A, tf.add(y, y, name='x_p_y'), name='A_dot')
+
+    range_out = io.StringIO()
+    with redirect_stdout(range_out):        
+        tf_dprint(z, depth_lower=1, depth_upper=5)
+
+    expected_out = textwrap.dedent('''
+    ... Tensor(MatMul):0,\tshape=[None, 1]\t"A_dot:0"
+    ... Op(MatMul)\t"A_dot"
+    ... |  Tensor(Placeholder):0,\tshape=[None, None]\t"A:0"
+    ... |  Tensor(Add):0,\tshape=[None, 1]\t"x_p_y:0"
+    ... |  |  Op(Add)\t"x_p_y"
+    ... |  |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
+    ... |  |  |  |  ...
+    ... |  |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
+    ... |  |  |  |  ...
+    ... ...
+    ''')
+
+    assert range_out.getvalue() == expected_out.lstrip()
+
+
+@pytest.mark.usefixtures("run_with_tensorflow")
+@pytest.mark.xfail(strict=True)
+@run_in_graph_mode
+def test_shape_index_type():
+    """Make sure graph indexing functions as expected."""
+
+    A = tf.compat.v1.placeholder('float', name='A',
+                                 shape=tf.TensorShape([None, None]))
+    x = tf.compat.v1.placeholder('float', name='x',
+                                 shape=tf.TensorShape([None, 1]))
+    y = tf.multiply(1.0, x, name='y')
+
+    z = tf.matmul(A, tf.add(y, y, name='x_p_y'), name='A_dot')
+
+    with pytest.raises(TypeError):
+        tf_dprint(z, depth_lower=3.2, depth_upper=6.2)
+
+
+@pytest.mark.usefixtures("run_with_tensorflow")
+@run_in_graph_mode
+def test_shape_index_with_repeated_elements():
+    """Make sure we can ascii/text print a TF graph
+    that is truncated and has repeated graph elements."""
+
+    A = tf.compat.v1.placeholder('float', name='A',
+                                 shape=tf.TensorShape([None, None]))
+    x = tf.compat.v1.placeholder('float', name='x',
+                                 shape=tf.TensorShape([None, 1]))
+    y = tf.multiply(1.0, x, name='y')
+
+    z = tf.matmul(A, tf.add(y, y, name='x_p_y'), name='A_dot')
+
+    std_out = io.StringIO()
+    with redirect_stdout(std_out):
+        tf_dprint(z, depth_lower=3, depth_upper=7)
+
+    expected_out = textwrap.dedent('''
+    ... Tensor(MatMul):0,\tshape=[None, 1]\t"A_dot:0"
+    ... Op(MatMul)\t"A_dot"
+    ... Tensor(Placeholder):0,\tshape=[None, None]\t"A:0"
+    ... Tensor(Add):0,\tshape=[None, 1]\t"x_p_y:0"
+    ... Op(Add)\t"x_p_y"
+    ... |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
+    ... |  |  Op(Mul)\t"y"
+    ... |  |  |  Tensor(Const):0,\tshape=[]\t"y/x:0"
+    ... |  |  |  Tensor(Placeholder):0,\tshape=[None, 1]\t"x:0"
+    ... |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
+    ... |  |  ...
+    ... ...
+    ''')
+
```

Comment:
This PR has a test that starts from a lower-bound that's at the minimum depth and an upper-bound that's before the maximum, as well as a test with a lower-bound greater than the minimum and an upper-bound at (or greater than) the maximum depth.
Now, all it needs is a test for a lower-bound above the minimum and an upper-bound below the maximum.

For example, `tf_dprint(z, depth_lower=3, depth_upper=5)`, which should have the following expected output:
```
... Tensor(Placeholder):0,\tshape=[None, None]\t"A:0"
... Tensor(Add):0,\tshape=[None, 1]\t"x_p_y:0"
... |  Op(Add)\t"x_p_y"
... |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
... |  |  |  ...
... |  |  Tensor(Mul):0,\tshape=[None, 1]\t"y:0"
... |  |  |  ...
```